apps/k9/server/src/main/java/net/discdd/app/k9/K9DDDAdapter.java

```diff
@@ -14,106 +37,186 @@
+        if (entity.isPresent()) {
+          String hashedPassword = entity.get().getPassword();
+          if (passwordEncoder.matches(parsedAdu.getPassword(), hashedPassword)) {
+            // update email clientId entry
+            clientToEmailRepository.updateClientId(clientId, parsedAdu.getEmail());
+            // create success ack
+            LoginAduAck ack = new LoginAduAck(parsedAdu.getEmail(), hashedPassword, true, null);
+            sendADUsStorage.addADU(clientId, APP_ID, ack.toByteArray(), -1);
+            return;
+          }
         }
+      }
+    }
 
-        List<AppDataUnit> dataListToReturn = new ArrayList<>();
-        List<ADU> aduListToReturn = new ArrayList<>();
-
-        try {
-            aduListToReturn = sendADUsStorage.getAppData(clientId, APP_ID);
-        } catch (IOException e) {
-            logger.log(SEVERE, "Error fetching ADUs to return for clientId: " + clientId, e);
+    // create error ack
+    LoginAduAck ack = new LoginAduAck(null, null, false, "Either not parsable, email doesn't exist, or password is incorrect.");
+    sendADUsStorage.addADU(clientId, APP_ID, ack.toByteArray(), -1);
+  }
+
+  private void processRegisterAdus(AppDataUnit adu, String clientId) throws IOException {
+    RegisterAdu parsedAdu = RegisterAdu.parseAdu(adu);
+    
+    if (parsedAdu != null) {
+      // generate emails
+      String[] emails = parsedAdu.generateEmails();
+      for (String email : emails) {
+        if (!clientToEmailRepository.existsById(email)) {
+          String hashedPassword = passwordEncoder.encode(parsedAdu.getPassword());
+          K9ClientIdToEmailMapping entity = new K9ClientIdToEmailMapping(email, clientId, hashedPassword);
+          clientToEmailRepository.save(entity);
+          RegisterAduAck ack = new RegisterAduAck(email, hashedPassword, true, null);
+          sendADUsStorage.addADU(clientId, APP_ID, ack.toByteArray(), -1);
+          return;
         }
-
-        try {
-            for (var adu : aduListToReturn) {
-                long aduId = adu.getADUId();
-                var data = sendADUsStorage.getADU(clientId, APP_ID, aduId);
-                dataListToReturn.add(
-                        AppDataUnit.newBuilder().setData(ByteString.copyFrom(data)).setAduId(aduId).build());
-            }
-        } catch (Exception e) {
-            logger.log(SEVERE, "Error while building response data for clientId: " + clientId, e);
+      }
+    }
+    
+    RegisterAduAck ack = new RegisterAduAck(null, null, false, "Either not parsable or all email combinations exist");
+    sendADUsStorage.addADU(clientId, APP_ID, ack.toByteArray(), -1);
+  }
+
+  // This method will parse the ToAddress from the mail ADUs and prepares ADUs for the respective
+  // client directories
+  private void processADUsToSend(AppDataUnit adu, String clientId) throws IOException {
+    String dataStr = new String(adu.getData().toByteArray());
```

Comment:
since the data could be quite big and we are just looking at the first few bytes, perhaps we should just convert the first few bytes.